### PR TITLE
STM32U0: provide support for ADC peripherals

### PR DESCRIPTION
**Pull request by djiatsaf-st**
This PR adds ADC peripherals on the stm32U0 boards.

### CODE REVIEW

**Review by erwango — COMMENTED** [review]
LGTM aside compliance checks

**Review by erwango — DISMISSED** [review]

**Review by decsny — DISMISSED** [review]

**Comment by erwango** [comment]
@djiatsaf-st Please rebase

**Review by anangl — DISMISSED** [review]

File: boards/st/nucleo_u031r8/nucleo_u031r8.dts [hunk 1]
```diff
@@ -81,3 +81,12 @@
 	ahb-prescaler = <1>;
 	apb1-prescaler = <1>;
 };
+
+&adc1 {
+	pinctrl-0 = <&adc1_in0_pc0 &adc1_in1_pc1>;
+	pinctrl-names = "default";
+	st,adc-clock-source = <ASYNC>;
```
Comment (gautierg-st) [inline boards/st/nucleo_u031r8/nucleo_u031r8.dts:88]: You set the clock source as asynchronous, but you didn't explicitly select which clock you wanted. It works but it is not good practice.
Either use SYNC, or define a secondary source clock.
To repeat on all three boards

File: drivers/adc/adc_stm32.c [hunk 2]
```diff
@@ -455,7 +455,8 @@ static void adc_stm32_disable(ADC_TypeDef *adc)
 	!defined(CONFIG_SOC_SERIES_STM32G0X) && \
 	!defined(CONFIG_SOC_SERIES_STM32L0X) && \
 	!defined(CONFIG_SOC_SERIES_STM32WBAX) && \
-	!defined(CONFIG_SOC_SERIES_STM32WLX)
+	!defined(CONFIG_SOC_SERIES_STM32WLX) && \
+	!defined(CONFIG_SOC_SERIES_STM32U0X)
```
Comment (gautierg-st) [inline drivers/adc/adc_stm32.c:459]: Keep the series in alphabetical order, this should go before WBA here. To repeat everywhere necessary.

File: tests/drivers/adc/adc_api/boards/nucleo_u083rc.overlay [hunk 3]
```diff
@@ -0,0 +1,25 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Copyright (c) 2024 STMicroelectronics
+ */
+
+/ {
+	zephyr,user {
+		/* adjust channel number according to pinmux in board.dts */
+		io-channels = <&adc1 0>;
```
Comment (gautierg-st) [inline tests/drivers/adc/adc_api/boards/nucleo_u083rc.overlay:10]: You defined two ADC pins in your dts. Use both in the overlays, that way we test sampling multiple channels in one go.
To repeat on your other overlay.

**Review by gautierg-st — CHANGES_REQUESTED** [review]

File: tests/drivers/adc/adc_api/boards/nucleo_u083rc.overlay [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+	zephyr,user {
+		/* adjust channel number according to pinmux in board.dts */
+		io-channels = <&adc1 0>, <&adc1 1>;
+	};
+};
+
+&adc1 {
+	st,adc-clock-source = <ASYNC>;
+	clocks = <&rcc STM32_CLOCK_BUS_APB1_2 0x00100000>,
+		 <&rcc STM32_SRC_HSI ADC_SEL(0)>;
```
Comment (gautierg-st) [inline tests/drivers/adc/adc_api/boards/nucleo_u083rc.overlay:17]: If you want to select HSI, this should be `ADC_SEL(2)`. `ADC_SEL(0)` will select the System Clock.

Besides, this clock configuration should be done directly in the board dts.

File: tests/drivers/adc/adc_api/boards/stm32u083c_dk.overlay [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+	zephyr,user {
+		/* adjust channel number according to pinmux in board.dts */
+		io-channels = <&adc1 0>, <&adc1 1>;
+	};
+};
+
+&adc1 {
+	st,adc-clock-source = <ASYNC>;
+	clocks = <&rcc STM32_CLOCK_BUS_APB1_2 0x00100000>,
+		 <&rcc STM32_SRC_HSI ADC_SEL(0)>;
```
Comment (gautierg-st) [inline tests/drivers/adc/adc_api/boards/stm32u083c_dk.overlay:17]: Same

**Review by gautierg-st — CHANGES_REQUESTED** [review]

File: tests/drivers/adc/adc_api/boards/nucleo_u083rc.overlay (hunk 4 above)
Comment (djiatsaf-st) [reply tests/drivers/adc/adc_api/boards/nucleo_u083rc.overlay:17]: Indeed, 
I did the tests with your suggestion,I don't know how I managed not to add these configurations after . 
Thank you.

**Review by gautierg-st — DISMISSED** [review]

**Review by erwango — DISMISSED** [review]

**Review by gautierg-st — DISMISSED** [review]

**Review by erwango — DISMISSED** [review]

**Comment by erwango** [comment]
@anangl PTAL

**Review by FRASTM — DISMISSED** [review]

File: tests/drivers/adc/adc_api/testcase.yaml [hunk 6]
```diff
@@ -7,6 +7,7 @@ tests:
   drivers.adc:
     depends_on: adc
     min_flash: 40
+    min_ram: 20
```
Comment (anangl) [inline tests/drivers/adc/adc_api/testcase.yaml:10]: This does not look right. The test itself does not require much RAM. Couldn't those problematic platforms be filtered out in some other way?

Comment (erwango) [reply tests/drivers/adc/adc_api/testcase.yaml:10]: I don't see the issue here. This is the usual way to filter out tests/samples based on memory footprint.

Here is the footprint report for a build on `nrf52840dk_nrf52840`
```
Memory region         Used Size  Region Size  %age Used
           FLASH:       86488 B         1 MB      8.25%
             RAM:       16624 B       256 KB      6.34%
        IDT_LIST:          0 GB        32 KB      0.00%

```
So RAM consumption for this test is above 16K while `nucleo_u031r8` available RAM is 12K.

Comment (anangl) [reply tests/drivers/adc/adc_api/testcase.yaml:10]: > Here is the footprint report for a build on `nrf52840dk_nrf52840`

Here's one for `nrf51dk/nrf51822`:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:       43280 B       256 KB     16.51%
             RAM:        8112 B        32 KB     24.76%
        IDT_LIST:          0 GB        32 KB      0.00%
```
It fits in 8 kB. And that's the point. It depends on the platform, not the test itself.

Comment (erwango) [reply tests/drivers/adc/adc_api/testcase.yaml:10]: Interesting. Fair enough, let's filter it in another way.

**Review by erwango — APPROVED** [review]

**Review by gautierg-st — APPROVED** [review]

**Review by anangl — APPROVED** [review]

**Review by FRASTM — APPROVED** [review]

**Pull request merged**